packages/studio-base/src/players/UserNodePlayer/index.ts

```
@@ -794,6 +799,28 @@ export default class UserNodePlayer implements Player {
         this.#setUserNodeDiagnostics(nodeRegistration.nodeId, []);
       }
     });
+
+    // If we have new topics after processing the node registrations we need to emit a new
+    // state to let downstream clients subscribe to newly available topics. This is
+    // necessary because we won't emit a new state otherwise if there are no other active
+    // subscriptions.
+    if (changedTopicsRequireEmitState && this.#playerState?.activeData) {
+      const newTopics = unionBy(
+        this.#playerState.activeData.topics,
+        this.#memoizedNodeTopics,
+        (top) => top.name,
+      );
+      const newDatatypes = this.#getDatatypes(
+        this.#playerState.activeData.datatypes,
+        this.#memoizedNodeDatatypes,
+      );
+      this.#playerState.activeData = {
+        ...this.#playerState.activeData,
+        topics: newTopics,
+        datatypes: newDatatypes,
+      };
+      this.#emitState().catch(console.error);
```

we use `void this.#emitState()` elsewhere in this file, would be nice to keep consistent

---

The `void promise` pattern seems undesirable to me. Shouldn't we at least log the error?

---

Actually I guess `void promise` is better here since it will throw if it fails instead of logging to console.